File: frontend/src/app/services/authorization.service.ts
import { Injectable, inject, computed } from '@angular/core';
import { AuthService } from './auth.service';

export type Role = 'ADMINISTRADOR' | 'OPERADOR';

export interface PermissaoModulo {
  id: string;
  rolesPermitidos: Role[];
  requerAutenticacao: boolean;
}

@Injectable({
  providedIn: 'root'
})
export class AuthorizationService {
  private readonly authService = inject(AuthService);

  readonly permissoesModulos = computed<Map<string, PermissaoModulo>>(() => {
    const permissoes = new Map<string, PermissaoModulo>();

    permissoes.set('cardapio', {
      id: 'cardapio',
      rolesPermitidos: ['ADMINISTRADOR'],
      requerAutenticacao: true
    });

    permissoes.set('lobby-pedidos', {
      id: 'lobby-pedidos',
      rolesPermitidos: ['ADMINISTRADOR', 'OPERADOR'],
      requerAutenticacao: true
    });

    permissoes.set('pedidos', {
      id: 'pedidos',
      rolesPermitidos: ['ADMINISTRADOR', 'OPERADOR'],
      requerAutenticacao: true
    });

    permissoes.set('fila-pedidos-mesa', {
      id: 'fila-pedidos-mesa',
      rolesPermitidos: ['ADMINISTRADOR', 'OPERADOR'],
      requerAutenticacao: true
    });

    permissoes.set('sessoes', {
      id: 'sessoes',
      rolesPermitidos: ['ADMINISTRADOR'],
      requerAutenticacao: true
    });

    permissoes.set('historico-sessoes', {
      id: 'historico-sessoes',
      rolesPermitidos: ['ADMINISTRADOR'],
      requerAutenticacao: true
    });

    permissoes.set('gestao-caixa', {
      id: 'gestao-caixa',
      rolesPermitidos: ['ADMINISTRADOR'],
      requerAutenticacao: true
    });

    permissoes.set('relatorios', {
      id: 'relatorios',
      rolesPermitidos: ['ADMINISTRADOR'],
      requerAutenticacao: true
    });

    permissoes.set('relatorio-financeiro', {
      id: 'relatorio-financeiro',
      rolesPermitidos: ['ADMINISTRADOR'],
      requerAutenticacao: true
    });

    permissoes.set('gestao-estoque', {
      id: 'gestao-estoque',
      rolesPermitidos: ['ADMINISTRADOR'],
      requerAutenticacao: true
    });

    permissoes.set('administracao', {
      id: 'administracao',
      rolesPermitidos: ['ADMINISTRADOR'],
      requerAutenticacao: true
    });

    return permissoes;
  });

  podeAcessarModulo(idModulo: string): boolean {
    if (!this.authService.estaAutenticado()) {
      return false;
    }

    const permissao = this.permissoesModulos().get(idModulo);
    if (!permissao) {
      return false;
    }

    const roleAtual = this.authService.getRoleAtual();
    if (!roleAtual) {
      return false;
    }

    return permissao.rolesPermitidos.includes(roleAtual as Role);
  }

  getRolesPermitidos(idModulo: string): Role[] {
    return this.permissoesModulos().get(idModulo)?.rolesPermitidos ?? [];
  }

  requerAdministrador(idModulo: string): boolean {
    const permissao = this.permissoesModulos().get(idModulo);
    if (!permissao) {
      return false;
    }
    return permissao.rolesPermitidos.length === 1 &&
      permissao.rolesPermitidos[0] === 'ADMINISTRADOR';
  }
}

File: autenticacao/src/main/java/com/sonecadelivery/autenticacao/domain/services/AuthorizationService.java
package com.sonecadelivery.autenticacao.domain.services;

import com.sonecadelivery.autenticacao.domain.entities.Role;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.stereotype.Service;

import java.util.Collection;
import java.util.Set;

@Service
public class AuthorizationService {
    
    private static final Set<String> MODULOS_ADMINISTRADOR = Set.of(
        "cardapio", "sessoes", "historico-sessoes", 
        "relatorios", "relatorio-financeiro", "administracao"
    );
    
    private static final Set<String> MODULOS_OPERADOR = Set.of(
        "pedidos", "lobby-pedidos"
    );
    
    public boolean podeAcessarModulo(String idModulo, Role role) {
        if (role == null) {
            return false;
        }
        
        if (role.isAdministrador()) {
            return true;
        }
        
        if (role.isOperador()) {
            return MODULOS_OPERADOR.contains(idModulo);
        }
        
        return false;
    }
    
    public boolean requerAdministrador(String idModulo) {
        return MODULOS_ADMINISTRADOR.contains(idModulo);
    }
    
    public boolean requerOperador(String idModulo) {
        return MODULOS_OPERADOR.contains(idModulo);
    }
    
    public Role getRoleAtual() {
        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
        
        if (authentication == null || !authentication.isAuthenticated()) {
            return null;
        }
        
        Collection<? extends GrantedAuthority> authorities = authentication.getAuthorities();
        
        if (authorities == null || authorities.isEmpty()) {
            return null;
        }
        
        String authority = authorities.iterator().next().getAuthority();
        
        try {
            return Role.fromAuthority(authority);
        } catch (IllegalArgumentException e) {
            return null;
        }
    }
    
    public boolean isAdministrador() {
        Role role = getRoleAtual();
        return role != null && role.isAdministrador();
    }
    
    public boolean isOperador() {
        Role role = getRoleAtual();
        return role != null && role.isOperador();
    }
    
    public boolean temRole(Role roleRequerido) {
        Role roleAtual = getRoleAtual();
        if (roleAtual == null || roleRequerido == null) {
            return false;
        }
        return roleAtual == roleRequerido;
    }
    
    public boolean temQualquerRole(Role... rolesRequeridos) {
        Role roleAtual = getRoleAtual();
        if (roleAtual == null || rolesRequeridos == null || rolesRequeridos.length == 0) {
            return false;
        }
        
        for (Role roleRequerido : rolesRequeridos) {
            if (roleAtual == roleRequerido) {
                return true;
            }
        }
        
        return false;
    }
}

